kernel/src/scan/log_replay.rs

```diff
@@ -220,24 +256,29 @@ impl LogReplayScanner {
         let mut visitor = AddRemoveDedupVisitor {
             seen: &mut self.seen,
             selection_vector,
+            logical_schema,
+            transform,
+            transforms: HashMap::new(),
             is_log_batch,
         };
         visitor.visit_rows_of(actions)?;
 
         // TODO: Teach expression eval to respect the selection vector we just computed so carefully!
         let selection_vector = visitor.selection_vector;
         let result = add_transform.evaluate(actions)?;
-        Ok((result, selection_vector))
+        Ok((result, selection_vector, visitor.transforms))
     }
 }
 
 /// Given an iterator of (engine_data, bool) tuples and a predicate, returns an iterator of
 /// `(engine_data, selection_vec)`. Each row that is selected in the returned `engine_data` _must_
 /// be processed to complete the scan. Non-selected rows _must_ be ignored. The boolean flag
 /// indicates whether the record batch is a log or checkpoint batch.
-pub fn scan_action_iter(
+pub(crate) fn scan_action_iter(
```

Comment:
Note this is a significant change as we not longer expose this function. In discussion so far we've agreed that it basically should never have been `pub`, and I just made a mistake when doing so. An engine should call `scan_data` which mostly just proxies to this, but doesn't expose internal details to the engine.

Open to discussion though.

Reply:
`pub(crate)` SGTM!